javascript/organizing_your_javascript_code/organizing_code_with_objects.md

```diff
@@ -0,0 +1,374 @@
+
+Feel free to copy this code in the console and experiment with it! If you're inclined, perhaps you could create a method to change the color of the light, as if it's one of those fancy RGB LEDs those gamer nerds and keyboard enthusiasts seem to like so much.
+
+Moving past physical items, we could also try to describe something a little bit more abstract like a game as an object. Since we've already explored Rock Paper Scissors in Foundations, let's use that as an example.
+
+A rock paper scissors game might involve a couple basic things:
+
+- Players' scores
+- The ability to play a round (and playing a round should update a player's score)
+
+And might also include a couple nice-to-haves
+
+- The ability to determine the current winning player
+- The ability to restart the game
+
+So, at it's most basic, an object that represents the game might look something like this (assuming we're playing against a computer player):
+
+```javascript
+const rps = {
+  playerScore: 0,
+  computerScore: 0,
+  playRound(playerChoice) {
+    // code to play the round... (and update the scores when a player wins)
+  }
+}
+```
+
+And if we fleshed it out, our object may come out to look something like this:
+
+```javascript
+const rps = {
+  playerScore: 0,
+  computerScore: 0,
+  playRound(playerChoice) {
+    const options = ['rock', 'paper', 'scissors']
+
+    // if an invalid choice is chosen, throw an error
+    if(!options.includes(playerChoice.toLowerCase())) {
+      throw new Error(`Expected 'rock', 'paper', or 'scissors', but got ${playerChoice}`)
+    }
+    
+    // get the computer's choice
+    const computerChoice = options[Math.floor(Math.random() * 3)]
+
+
+    // determine the winner, apply points if necessary, and return who won
+    if(playerChoice.toLowerCase() === computerChoice) {
+      return "tie"
+    } else if(
+      (playerChoice === 'rock' && computerChoice === 'scissors') ||
+      (playerChoice === 'paper' && computerChoice === 'rock') ||
+      (playerChoice === 'scissors' && computerChoice === 'paper')
+    ) {
+      this.playerScore++
+      return "player"
+    } else {
+      this.computerScore++
+      return 'computer'
+    }
+  },
+  getWinningPlayer() {
+    if(this.playerScore === this.computerScore) {
+      return 'tie'
+    } else if (this.playerScore > this.computerScore) {
+      return 'player'
+    } else {
+      return 'computer'
+    }
+  },
+  reset() {
+    this.playerScore = 0;
+    this.computerScore = 0;
+  }
+}
+
+rps.playRound('rock') // returns 'player' if we win...
+rps.playerScore       // ...and our score would have increased
+
+// We also have the ability to check the winner and reset the game at any time
+rps.getWinningPlayer() // 'player', if we won above round
+rps.reset()
+```
```

Comment:
Probably more of a nit but I'm not sure we really need such a big code block here? I'm just thinking more about how such a big code block can have the psychological effect of "oh god" (like a red herring) when the code within the methods doesn't actually matter for the lesson's purpose. We could replace the methods contents with a short comment. Those, along with the method names as well as the "driver script" underneath, I believe should be sufficient to get the point across (while allowing learners to focus more on that point than trying to parse the RPS code you wrote).

Reply:
I do like the full, working code example personally, but this can happen.

I'm thinking I can just replace the implementation with pseudo-code as a middle ground:
```js
const rps = {
  playerScore: 0,
  computerScore: 0,
  playRound(playerChoice) {
    /* 
      1. if an invalid choice is chosen, throw an error
      
      2. get the computer's choice
      
      3. determine the winner
        - apply points if necessary
        - return who won ('player', 'computer', or 'tie')
     */
  },
  getWinningPlayer() {
    // return the player with the most points ('player', 'computer', or 'tie')
  },
  reset() {
    // reset both player's scores to 0
  }
}

rps.playRound('rock') // returns 'player' if we win...
rps.playerScore       // ...and our score would have increased

// We also have the ability to check the winner and reset the game at any time
rps.getWinningPlayer() // 'player', if we won above round
rps.reset()
```

It might be a nice little exercise to suggest to learners that they can try and fill in the code and get it working if they want.

My concerns with doing it this way is that it makes it hard to extract particular parts of the `playRound` function to exemplify how they can be pulled into private methods.

A workaround I can think of is making the filling in of the above pseudocode an exercise learners are directed to do, then approach private properties. That way they can learn how to break their functions down in an active way, while still having a nice interface on their object others are meant to use.

I'm not sure how great of an idea that is- it relies on the learner working within the framework provided, but I believe I can make it work. It helps a lot that I decided to use an example learners should be very familiar with.

Reply:
The thing I'm concerned with is the scope of the lesson given where the learner will be, and putting the right weight on the different concepts the lesson wants to explore.

When it comes to the concept of private properties, I'm not convinced this lesson needs to be the one where we really dive into that concept and demonstrating its uses. Privacy already gets explored a lot more in the factory functions lesson, and revisited in the classes lesson (and from my observations, by that point, many learners seem to have grasped the concept of private properties and their use).

So my feeling is that this lesson probably won't need much more diving into private properties than an outline of the idea and that we'll explore more ways to implement and make use of privacy later on.

Therefore when it comes to this RPS example, I don't think we actually need the functions' internal code to make a sufficient point. Including that code I feel puts far too much weight on this part of the lesson which could take weight away from the other parts that may actually need more of that weight.

An example of how privacy could tie into the RPS code with the truncated version is just outlining how you may want to extract the logic for getting human and computer choices from `playRound` to their own functions (we don't need to show the full code for that since that the idea should be intuitive by now). Then since those choice functions are only called internally by `playRound` and not externally, they can be treated as "private". We won't need to show implementation details of the actual functions to demonstrate that idea, and to me it also puts most of the weight for this section on that idea instead of the actual RPS code itself.

Reply:
> Privacy already gets explored a lot more in the factory functions lesson, and revisited in the classes lesson (and from my observations, by that point, many learners seem to have grasped the concept of private properties and their use).

I just re-read this lesson, and realized it goes into this more in-depth than I'd thought, so I won't argue much about that.

I see your point about a full code examples, and I've just made some changes to reflect that. I think we're on a similar page on how it can be rectified.

I really wanted to include private properties/methods in this lesson, since it is a overarching concept that will come up regardless of one's object-creation method, so it feels a little hard for me to let go of. 

Let me know what you think of the most recent changes, we can continue discussing it here.